src/api/accounts.rs

```diff
@@ -165,74 +186,172 @@ mod accounts_signing {
+            stream.append(&chain_id);
+
+            self.rlp_append_legacy(&mut stream);
+            self.rlp_append_access_list(&mut stream);
+
+            if let Some(signature) = signature {
+                self.rlp_append_signature(&mut stream, signature);
             }
-            rlp.append(&self.value);
-            rlp.append(&self.data);
-            rlp.append(&chain_id);
-            rlp.append(&0u8);
-            rlp.append(&0u8);
+
+            stream
         }
 
-        /// RLP encode a signed transaction with the specified signature.
-        fn rlp_append_signed(&self, rlp: &mut RlpStream, signature: &Signature) {
-            rlp.begin_list(9);
-            rlp.append(&self.nonce);
-            rlp.append(&self.gas_price);
-            rlp.append(&self.gas);
+        fn encode_eip1559_payload(&self, chain_id: u64, signature: Option<&Signature>) -> RlpStream {
+            let mut stream = RlpStream::new();
+
+            let list_size = if signature.is_some() { 12 } else { 9 };
+            stream.begin_list(list_size);
+
+            // append chain_id. from EIP-2930: chainId is defined to be an integer of arbitrary size.
+            stream.append(&chain_id);
+
+            stream.append(&self.nonce);
+            stream.append(&self.max_priority_fee_per_gas);
+            stream.append(&self.gas_price);
+            stream.append(&self.gas);
             if let Some(to) = self.to {
-                rlp.append(&to);
+                stream.append(&to);
             } else {
-                rlp.append(&"");
+                stream.append(&"");
+            }
+            stream.append(&self.value);
+            stream.append(&self.data);
+
+            self.rlp_append_access_list(&mut stream);
+
+            if let Some(signature) = signature {
+                self.rlp_append_signature(&mut stream, signature);
+            }
+
+            stream
+        }
+
+        fn rlp_append_signature(&self, stream: &mut RlpStream, signature: &Signature) -> () {
+            stream.append(&signature.v);
+            stream.append(&U256::from_big_endian(signature.r.as_bytes()));
+            stream.append(&U256::from_big_endian(signature.s.as_bytes()));
+        }
+
+        fn rlp_append_access_list(&self, stream: &mut RlpStream) -> () {
+            stream.begin_list(self.access_list.len());
+            for access in self.access_list.iter() {
+                stream.begin_list(2);
+                stream.append(&access.address);
+                stream.begin_list(access.storage_keys.len());
+                for storage_key in access.storage_keys.iter() {
+                    stream.append(storage_key);
+                }
+            }
+        }
+
+        fn encode(&self, chain_id: u64, signature: Option<&Signature>) -> Vec<u8> {
+            match self.transaction_type.map(|t| t.as_u64()) {
+                // type 0: Legacy
```

Comment:
I'd prefer these types to be either an `enum` (converted into `u8/u16/u64`) or a set of constants.

Reply:
Added constants.